File: src/test/java/leetcode/LC003LongestSubstringWithoutRepeatingCharacters.java
package leetcode;

import java.util.HashSet;
import java.util.Set;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */
public class LC003LongestSubstringWithoutRepeatingCharacters {
    public class Solution {
        public int lengthOfLongestSubstring(String s) {
            Set<Character> seen = new HashSet<>();
            int longest = 0;
            int count = 0;
            for (int i = 0; i < s.length(); i++) {
                Character found = s.charAt(i);
                while(count > 0 && seen.contains(found)) {
                    seen.remove(s.charAt(i - count));
                    count--;
                }
                count++;
                seen.add(found);
                longest = Math.max(longest, count);
            }
            return longest;
        }
    }

    @Test
    public void test_abcabcbb() throws Exception {
        assertEquals(3, new Solution().lengthOfLongestSubstring("abcabcbb"));
    }

    @Test
    public void test_bbbbb() throws Exception {
        assertEquals(1, new Solution().lengthOfLongestSubstring("bbbbb"));
    }

    @Test
    public void test_dvdf() throws Exception {
        assertEquals(3, new Solution().lengthOfLongestSubstring("dvdf"));
    }
}

File: algorithms/code/leetcode/lc003_longest_substring_without_repeating_characters/lc003_longest_substring_without_repeating_characters.py
#!/usr/bin/env python3
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

import unittest

from typing import Set


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen: Set[str] = set()
        longest = count = 0
        for i, found in enumerate(s):
            while count > 0 and found in seen:
                seen.remove(s[i - count])
                count -= 1
            count += 1
            seen.add(found)
            longest = max(longest, count)
        return longest


class TestCode(unittest.TestCase):
    def test_abcabcbb(self):
        self.assertEqual(Solution().lengthOfLongestSubstring('abcabcbb'), 3)

    def test_bbbbb(self):
        self.assertEqual(Solution().lengthOfLongestSubstring('bbbbb'), 1)

    def test_dvdf(self):
        self.assertEqual(Solution().lengthOfLongestSubstring('dvdf'), 3)
